org.mwc.debrief.legacy/src/Debrief/ReaderWriter/Replay/ImportReplay.java

```diff
@@ -1275,72 +1395,128 @@ final private String getThisSymProperty(final String symbology,
     }
     return null;
   }
+  
+  private final static void addElementsToExistingLayer(final Layer layerToAddTo,final Editable l)
+  {
+    Enumeration<Editable> tempElements = ((Layer)l).elements();
+    while(tempElements.hasMoreElements()) {
+      Editable elem = tempElements.nextElement();
+      layerToAddTo.add(elem);
+    }
+  }
+  public final static void injectContent(final Layers from,final Layers destination,final boolean performAdd) {
+    Enumeration<Editable> tempElements = from.elements();
+    //now add to the plot's layers object
+    while(tempElements.hasMoreElements()) {
+      Editable l = tempElements.nextElement();
+      Layer existingLayer = destination.findLayer(l.getName());
+      if(existingLayer==null) {
+        Layer layerToAdd = (Layer)l;
+        if(performAdd) {
+          destination.addThisLayer(layerToAdd);
+        }
+        else {
+          destination.removeThisLayer(layerToAdd);
+        }
+      }
+      else {
+        //get elements in the templayer for an already existing layer and add to it.
+        if(performAdd) {
+          addElementsToExistingLayer(existingLayer,l);
+        }
+        else {
+          if(l instanceof Layer) {
+            //dont directly delete a layer, may be there were some elements previously existing
+            Enumeration<Editable> elements = ((Layer)l).elements();
+            Layer destLayer = destination.findLayer(l.getName());
+            while(elements.hasMoreElements()) {
+              destination.removeThisEditable(destLayer, elements.nextElement());
+            }
+            //in case no more elements are left delete the layer
+            if(destLayer.elements()==null || !destLayer.elements().hasMoreElements()) {
+              destination.removeThisLayer(destLayer);
+            }
+          }
+        }
+      }
+    }
+    destination.fireExtended();
+  }
+
+  public final void importThis(final String text, final int numLines) {
+    final InputStream stream = new ByteArrayInputStream(text.getBytes());
+    importRep(null,stream,numLines);
+  }
 
+  @Override
+  public final void importThis(final String fName, InputStream is) {
+    final int numLines = countLinesFor(fName);
+    importRep(fName,is,numLines);
+  }
   /**
    * import data from this stream
    */
-  @Override
-  public final void importThis(final String fName, final InputStream is)
+  private final void importRep(final String fName,final InputStream is,final int numLines)
   {
     // declare linecounter
     int lineCounter = 0;
-
-    final int numLines = countLinesFor(fName);
-
-    final Reader reader = new InputStreamReader(is);
-    final BufferedReader br = new ReaderMonitor(reader, numLines, fName);
+    final Reader reader;
+    BufferedReader br = null;
     String thisLine = null;
     try
     {
```

Comment:
@ayeshare - in the following block, we duplicate the ReaderMonitor constructor.  This introduces a _slight_ fragility in having to track that we're correctly using reader and numLines, if the API changes at all. Just having the single use of the constructor keeps things simpler.

So, it would be _slightly_ better code if we just had the one constructor:
````
        final String nameToUse;
        if(fName != null)
        {
          nameToUse = fName;
        }
        else
        {
          nameToUse = "Pasted REP content";
        }
        br = new ReaderMonitor(reader, numLines, nameToUse);
````

A more terse version would be:
````
        final String nameToUse = fName != null ? fName : "Pasted REP content";
        br = new ReaderMonitor(reader, numLines, nameToUse);
````

Reply:
Note: I've also expanded the text we use in the filename param, when we're pasting from REP.